test/cql-pytest/test_compaction_tombstone_gc.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+from util import new_materialized_view, new_test_table
+import nodetool
+import time
+
+def test_tombstone_gc_with_conflict_in_memtable(scylla_only, cql, test_keyspace):
+    """
+    Regression test for fixed https://github.com/scylladb/scylladb/issues/20423
+    """
+    schema = "k int, v int, primary key (k, v)"
+    with new_test_table(cql, test_keyspace, schema, extra="with gc_grace_seconds = 0") as table:
+        with nodetool.no_autocompaction_context(cql, table):
+            nodetool.setlogginglevel(cql, "compaction", "trace")
+            # Insert initial data into the base table:
+            # Row 1 is expected to be garbage collected after expiration
+            cql.execute(f"insert into {table} (k, v) values (1, 1) using timestamp 1 and ttl 1")
+            cql.execute(f"insert into {table} (k, v) values (1, 2) using timestamp 2")
+            nodetool.flush(cql, table)
+            # Delete row 2, this deletion is expected to be kept
+            # when we insert backdated live data into the memtable
+            cql.execute(f"delete from {table} using timestamp 3 where k=1 and v=2")
+            # Flush all tables explicitly now, since this is skipped in the test on purpose by next major compaction.
+            nodetool.flush_all(cql)
+
+            # let gc_grace_seconds and ttl pass
+            time.sleep(2)
+            # Re-insert backdated data into the memtable.  It should inhibit tombstone_gc
```

Comment:
I'm worried that tombstone_gc was actually inhibited for a different reason, and if that's the case this test doesn't test your patch. For example above I asked why this tombstone wasn't immediately gc'ed when it was created and guessed maybe it's because it's newer than that "reordering grace period". If this is the case, then doesn't the same reordering grace period protect this tombstone in this place in the code as well?

Reply:
maybe it's the commitlog min_gc time that's the problem.
This is why the sleep is needed before inserting to the memtable.
Insert immediately and then wait for 2 seconds inhibits tombstone_gc because of the commitlog min_gc time.